File: backend/encryption_service.py
"""
Encryption service for the Journal Notes backend.

This module provides encryption and decryption capabilities for note data
using industry-standard cryptographic algorithms.
"""

import os
import base64
import hashlib
from typing import Union
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend


class EncryptionService:
    """Service for encrypting and decrypting data using AES-256-GCM."""
    
    def __init__(self):
        self.backend = default_backend()
    
    def _derive_key(self, password: str, salt: bytes) -> bytes:
        """
        Derive an encryption key from a password using PBKDF2.
        
        Args:
            password: The user's password
            salt: Random salt bytes
            
        Returns:
            32-byte encryption key
        """
        kdf = PBKDF2HMAC(
            algorithm=hashes.SHA256(),
            length=32,  # 256 bits
            salt=salt,
            iterations=100000,
            backend=self.backend
        )
        return kdf.derive(password.encode('utf-8'))
    
    def encrypt(self, data: str, password: str) -> str:
        """
        Encrypt data using AES-256-GCM.
        
        Args:
            data: The plaintext data to encrypt
            password: The encryption password
            
        Returns:
            Base64-encoded encrypted data (salt + nonce + ciphertext + tag)
        """
        # Generate random salt and nonce
        salt = os.urandom(16)  # 128 bits
        nonce = os.urandom(12)  # 96 bits for GCM
        
        # Derive key from password
        key = self._derive_key(password, salt)
        
        # Encrypt the data
        cipher = Cipher(algorithms.AES(key), modes.GCM(nonce), backend=self.backend)
        encryptor = cipher.encryptor()
        ciphertext = encryptor.update(data.encode('utf-8')) + encryptor.finalize()
        
        # Combine salt + nonce + ciphertext + tag
        encrypted_data = salt + nonce + ciphertext + encryptor.tag
        
        # Return base64-encoded result
        return base64.b64encode(encrypted_data).decode('utf-8')
    
    def decrypt(self, encrypted_data: str, password: str) -> str:
        """
        Decrypt data using AES-256-GCM.
        
        Args:
            encrypted_data: Base64-encoded encrypted data
            password: The decryption password
            
        Returns:
            The decrypted plaintext data
            
        Raises:
            ValueError: If decryption fails (wrong password or corrupted data)
        """
        try:
            # Decode from base64
            data = base64.b64decode(encrypted_data.encode('utf-8'))
            
            # Extract components
            salt = data[:16]
            nonce = data[16:28]
            ciphertext = data[28:-16]
            tag = data[-16:]
            
            # Derive key from password
            key = self._derive_key(password, salt)
            
            # Decrypt the data
            cipher = Cipher(algorithms.AES(key), modes.GCM(nonce, tag), backend=self.backend)
            decryptor = cipher.decryptor()
            plaintext = decryptor.update(ciphertext) + decryptor.finalize()
            
            return plaintext.decode('utf-8')
            
        except Exception as e:
            raise ValueError(f"Decryption failed: {str(e)}") from e

File: frontend/src/services/EncryptionService.ts
/**
 * EncryptionService - Client-side encryption/decryption using Web Crypto API
 * 
 * Implements AES-256-GCM encryption with PBKDF2 key derivation for secure
 * client-side data protection. All data is encrypted before storage.
 */

export class EncryptionService {
  private static readonly ALGORITHM = 'AES-GCM';
  private static readonly KEY_LENGTH = 256;
  private static readonly PBKDF2_ITERATIONS = 100000;
  private static readonly SALT_LENGTH = 16; // 128 bits
  private static readonly IV_LENGTH = 12;   // 96 bits for GCM

  /**
   * Derives a cryptographic key from a password using PBKDF2
   * @param password - User password
   * @param salt - Random salt for key derivation
   * @returns Promise<CryptoKey> - Derived encryption key
   */
  private async deriveKey(password: string, salt: Uint8Array): Promise<CryptoKey> {
    // Import the password as a key for PBKDF2
    const passwordKey = await crypto.subtle.importKey(
      'raw',
      new TextEncoder().encode(password),
      'PBKDF2',
      false,
      ['deriveKey']
    );
    
    // Derive the actual encryption key using PBKDF2
    return crypto.subtle.deriveKey(
      {
        name: 'PBKDF2',
        salt: salt,
        iterations: EncryptionService.PBKDF2_ITERATIONS,
        hash: 'SHA-256'
      },
      passwordKey,
      { 
        name: EncryptionService.ALGORITHM, 
        length: EncryptionService.KEY_LENGTH 
      },
      false,
      ['encrypt', 'decrypt']
    );
  }

  /**
   * Encrypts data using AES-256-GCM
   * @param data - Plain text data to encrypt
   * @param password - User password for key derivation
   * @returns Promise<string> - Base64 encoded encrypted data (salt + iv + ciphertext)
   */
  async encrypt(data: string, password: string): Promise<string> {
    if (!data) {
      throw new Error('Data cannot be empty');
    }
    
    if (!password) {
      throw new Error('Password cannot be empty');
    }

    // Generate random salt and IV
    const salt = crypto.getRandomValues(new Uint8Array(EncryptionService.SALT_LENGTH));
    const iv = crypto.getRandomValues(new Uint8Array(EncryptionService.IV_LENGTH));
    
    // Derive encryption key
    const key = await this.deriveKey(password, salt);
    
    // Encrypt the data
    const encodedData = new TextEncoder().encode(data);
    const encrypted = await crypto.subtle.encrypt(
      { 
        name: EncryptionService.ALGORITHM, 
        iv: iv 
      },
      key,
      encodedData
    );
    
    // Combine salt + iv + encrypted data
    const combined = new Uint8Array(salt.length + iv.length + encrypted.byteLength);
    combined.set(salt, 0);
    combined.set(iv, salt.length);
    combined.set(new Uint8Array(encrypted), salt.length + iv.length);
    
    // Return as base64 string
    return this.arrayBufferToBase64(combined);
  }

  /**
   * Decrypts data using AES-256-GCM
   * @param encryptedData - Base64 encoded encrypted data
   * @param password - User password for key derivation
   * @returns Promise<string> - Decrypted plain text data
   */
  async decrypt(encryptedData: string, password: string): Promise<string> {
    if (!encryptedData) {
      throw new Error('Encrypted data cannot be empty');
    }
    
    if (!password) {
      throw new Error('Password cannot be empty');
    }

    try {
      // Decode from base64
      const combined = this.base64ToArrayBuffer(encryptedData);
      
      // Extract salt, iv, and encrypted data
      const salt = combined.slice(0, EncryptionService.SALT_LENGTH);
      const iv = combined.slice(
        EncryptionService.SALT_LENGTH, 
        EncryptionService.SALT_LENGTH + EncryptionService.IV_LENGTH
      );
      const encrypted = combined.slice(EncryptionService.SALT_LENGTH + EncryptionService.IV_LENGTH);
      
      // Derive decryption key
      const key = await this.deriveKey(password, salt);
      
      // Decrypt the data
      const decrypted = await crypto.subtle.decrypt(
        { 
          name: EncryptionService.ALGORITHM, 
          iv: iv 
        },
        key,
        encrypted
      );
      
      // Return as string
      return new TextDecoder().decode(decrypted);
    } catch (error) {
      // Re-throw with more specific error message
      if (error instanceof Error && error.name === 'OperationError') {
        throw new Error('Decryption failed: Invalid password or corrupted data');
      }
      throw error;
    }
  }

  /**
   * Converts ArrayBuffer to base64 string
   * @param buffer - ArrayBuffer to convert
   * @returns string - Base64 encoded string
   */
  private arrayBufferToBase64(buffer: Uint8Array): string {
    let binary = '';
    const bytes = new Uint8Array(buffer);
    const len = bytes.byteLength;
    for (let i = 0; i < len; i++) {
      binary += String.fromCharCode(bytes[i]);
    }
    return btoa(binary);
  }

  /**
   * Converts base64 string to ArrayBuffer
   * @param base64 - Base64 encoded string
   * @returns Uint8Array - Decoded array buffer
   */
  private base64ToArrayBuffer(base64: string): Uint8Array {
    const binary = atob(base64);
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) {
      bytes[i] = binary.charCodeAt(i);
    }
    return bytes;
  }

  /**
   * Generates a random salt for key derivation
   * @returns Uint8Array - Random salt
   */
  generateSalt(): Uint8Array {
    return crypto.getRandomValues(new Uint8Array(EncryptionService.SALT_LENGTH));
  }

  /**
   * Generates a random IV for encryption
   * @returns Uint8Array - Random IV
   */
  generateIV(): Uint8Array {
    return crypto.getRandomValues(new Uint8Array(EncryptionService.IV_LENGTH));
  }
}